crates/matrix-sdk/src/latest_events/latest_event.rs

```diff
@@ -150,6 +169,332 @@ impl LatestEventValue {
+/// all the following local events must also be marked as wedged. And vice
+/// versa, when the send queue is able to send an event again, all the following
+/// local events must be marked as unwedged.
+///
+/// This type isolates a couple of methods designed to manage these specific
+/// behaviours.
+#[derive(Debug)]
+struct LatestEventValuesForLocalEvents {
+    buffer: Vec<(OwnedTransactionId, LatestEventValue)>,
+}
+
+impl LatestEventValuesForLocalEvents {
+    /// Create a new [`LatestEventValuesForLocalEvents`].
+    fn new() -> Self {
+        Self { buffer: Vec::with_capacity(2) }
+    }
+
+    /// Get the last [`LatestEventValue`].
+    fn last(&self) -> Option<&LatestEventValue> {
+        self.buffer.last().map(|(_, value)| value)
+    }
+
+    /// Find the position of the [`LatestEventValue`] matching `transaction_id`.
+    fn position(&self, transaction_id: &TransactionId) -> Option<usize> {
+        self.buffer
+            .iter()
+            .position(|(transaction_id_candidate, _)| transaction_id == transaction_id_candidate)
+    }
+
+    /// Push a new [`LatestEventValue`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if `value` is not of kind [`LatestEventValue::LocalIsSending`] or
+    /// [`LatestEventValue::LocalIsWedged`].
+    fn push(&mut self, transaction_id: OwnedTransactionId, value: LatestEventValue) {
+        assert!(
+            matches!(
+                value,
+                LatestEventValue::LocalIsSending(_) | LatestEventValue::LocalIsWedged(_)
+            ),
+            "`value` must be either `LocalIsSending` or `LocalIsWedged`"
+        );
+
+        self.buffer.push((transaction_id, value));
+    }
+
+    /// Replace the [`LatestEventKind`] of the [`LatestEventValue`] at position
+    /// `position`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if:
+    /// - `position` is strictly greater than buffer's length,
+    /// - the [`LatestEventValue`] is not of kind
+    ///   [`LatestEventValue::LocalIsSending`] or
+    ///   [`LatestEventValue::LocalIsWedged`].
+    fn replace_kind(&mut self, position: usize, new_kind: LatestEventKind) {
+        let (_, value) = self.buffer.get_mut(position).expect("`position` must be valid");
+
+        match value {
+            LatestEventValue::LocalIsSending(kind) => *kind = new_kind,
+            LatestEventValue::LocalIsWedged(kind) => *kind = new_kind,
+            _ => panic!("`value` must be either `LocalIsSending` or `LocalIsWedged`"),
+        }
+    }
+
+    /// Remove the [`LatestEventValue`] at position `position`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `position` is strictly greater than buffer's length.
+    fn remove(&mut self, position: usize) -> (OwnedTransactionId, LatestEventValue) {
+        self.buffer.remove(position)
+    }
+
+    /// Mark the `LatestEventValue` matching `transaction_id`, and all the
+    /// following values, as wedged.
+    fn wedged_from(&mut self, transaction_id: &TransactionId) {
```

Comment:
It's weird to have an effectful method with a name that's in the passive form. Consider renaming `mark_wedged_from`?